Rest arguments don't work with mapped types

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  `3.1.0`, `3.1.0-dev.20180810`

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**

mapped types, rest parameters

**Code**

```ts
interface A {
    emit(event: string, ...args: any[]): boolean;
}

type Args<F> = F extends (...args: infer A) => void ? A : never;

type EventMap = Record<string, Function>;

interface B<M extends EventMap> extends A {
    emit<Event extends keyof M>(event: Event, ...args: Args<M[Event]>): boolean;
}
```

**Expected behavior:**

No errors.

**Actual behavior:**

Compile error:

```
t.ts:9:11 - error TS2430: Interface 'B<M>' incorrectly extends interface 'A'.
  Types of property 'emit' are incompatible.
    Type '<Event extends keyof M>(event: Event, ...args: Args<M[Event]>) => boolean' is not assignable to type '(event: string, ...args: any[]) => boolean'.

9 interface B<M extends EventMap> extends A {
     
```

If I add any overload for `emit` function error is gone, e.g.:

```diff
interface A {
    emit(event: string, ...args: any[]): boolean;
}

type Args<F> = F extends (...args: infer A) => void ? A : never;

type EventMap = Record<string, Function>;

interface B<M extends EventMap> extends A {
    emit<Event extends keyof M>(event: Event, ...args: Args<M[Event]>): boolean;
+   emit(event: 'foo'): boolean;
}
```

***

Also, if change declaration of `EventMap` type to

```ts
type EventMap = Record<string, (...args: any[]) => void>;
```

then most common type for interface `B`

```ts
interface B<M extends EventMap> extends A {
    emit<Event extends keyof M>(event: Event, ...args: Args<M[Event]>): boolean;
}
```

is

```ts
interface C extends A {
    emit(event: string, ...args: Args<(...args: any[]) => void>): boolean;
}
```

But latest interface compiles without error: [playground link](http://www.typescriptlang.org/play/index.html#src=interface%20A%20%7B%0A%20%20%20%20emit(event%3A%20string%2C%20...args%3A%20any%5B%5D)%3A%20boolean%3B%0A%7D%0A%0Atype%20Args%3CF%3E%20%3D%20F%20extends%20(...args%3A%20infer%20A)%20%3D%3E%20void%20%3F%20A%20%3A%20never%3B%0A%0Atype%20EventMap%20%3D%20Record%3Cstring%2C%20(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E%3B%0A%0Ainterface%20C%20extends%20A%20%7B%0A%20%20%20%20emit(event%3A%20string%2C%20...args%3A%20Args%3C(...args%3A%20any%5B%5D)%20%3D%3E%20void%3E)%3A%20boolean%3B%0A%7D)


**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> [link](http://www.typescriptlang.org/play/index.html#src=interface%20A%20%7B%0A%20%20%20%20emit(event%3A%20string%2C%20...args%3A%20any%5B%5D)%3A%20boolean%3B%0A%7D%0A%0Atype%20Args%3CF%3E%20%3D%20F%20extends%20(...args%3A%20infer%20A)%20%3D%3E%20void%20%3F%20A%20%3A%20never%3B%0A%0Atype%20EventMap%20%3D%20Record%3Cstring%2C%20Function%3E%3B%0A%0Ainterface%20B%3CM%20extends%20EventMap%3E%20extends%20A%20%7B%0A%20%20%20%20emit%3CEvent%20extends%20keyof%20M%3E(event%3A%20Event%2C%20...args%3A%20Args%3CM%5BEvent%5D%3E)%3A%20boolean%3B%0A%7D%0A)

**Related Issues:** https://github.com/Microsoft/TypeScript/issues/26414 (Previous issue was closed without answering on questions).

--- a/tests/baselines/reference/eventEmitterPatternWithRecordOfFunction.js
+++ b/tests/baselines/reference/eventEmitterPatternWithRecordOfFunction.js
@@ -1,0 +1,15 @@
+//// [eventEmitterPatternWithRecordOfFunction.ts]
+interface A {
+    emit(event: string, ...args: any[]): boolean;
+}
+
+type Args<F> = F extends (...args: infer A) => void ? A : never;
+
+type EventMap = Record<string, Function>;
+
+interface B<M extends EventMap> extends A {
+    emit<Event extends keyof M>(event: Event, ...args: Args<M[Event]>): boolean;
+}
+
+//// [eventEmitterPatternWithRecordOfFunction.js]
+"use strict";

--- a/tests/baselines/reference/eventEmitterPatternWithRecordOfFunction.symbols
+++ b/tests/baselines/reference/eventEmitterPatternWithRecordOfFunction.symbols
@@ -1,0 +1,40 @@
+=== tests/cases/compiler/eventEmitterPatternWithRecordOfFunction.ts ===
+interface A {
+>A : Symbol(A, Decl(eventEmitterPatternWithRecordOfFunction.ts, 0, 0))
+
+    emit(event: string, ...args: any[]): boolean;
+>emit : Symbol(A.emit, Decl(eventEmitterPatternWithRecordOfFunction.ts, 0, 13))
+>event : Symbol(event, Decl(eventEmitterPatternWithRecordOfFunction.ts, 1, 9))
+>args : Symbol(args, Decl(eventEmitterPatternWithRecordOfFunction.ts, 1, 23))
+}
+
+type Args<F> = F extends (...args: infer A) => void ? A : never;
+>Args : Symbol(Args, Decl(eventEmitterPatternWithRecordOfFunction.ts, 2, 1))
+>F : Symbol(F, Decl(eventEmitterPatternWithRecordOfFunction.ts, 4, 10))
+>F : Symbol(F, Decl(eventEmitterPatternWithRecordOfFunction.ts, 4, 10))
+>args : Symbol(args, Decl(eventEmitterPatternWithRecordOfFunction.ts, 4, 26))
+>A : Symbol(A, Decl(eventEmitterPatternWithRecordOfFunction.ts, 4, 40))
+>A : Symbol(A, Decl(eventEmitterPatternWithRecordOfFunction.ts, 4, 40))
+
+type EventMap = Record<string, Function>;
+>EventMap : Symbol(EventMap, Decl(eventEmitterPatternWithRecordOfFunction.ts, 4, 64))
+>Record : Symbol(Record, Decl(lib.es5.d.ts, --, --))
+>Function : Symbol(Function, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --))
+
+interface B<M extends EventMap> extends A {
+>B : Symbol(B, Decl(eventEmitterPatternWithRecordOfFunction.ts, 6, 41))
+>M : Symbol(M, Decl(eventEmitterPatternWithRecordOfFunction.ts, 8, 12))
+>EventMap : Symbol(EventMap, Decl(eventEmitterPatternWithRecordOfFunction.ts, 4, 64))
+>A : Symbol(A, Decl(eventEmitterPatternWithRecordOfFunction.ts, 0, 0))
+
+    emit<Event extends keyof M>(event: Event, ...args: Args<M[Event]>): boolean;
+>emit : Symbol(B.emit, Decl(eventEmitterPatternWithRecordOfFunction.ts, 8, 43))
+>Event : Symbol(Event, Decl(eventEmitterPatternWithRecordOfFunction.ts, 9, 9))
+>M : Symbol(M, Decl(eventEmitterPatternWithRecordOfFunction.ts, 8, 12))
+>event : Symbol(event, Decl(eventEmitterPatternWithRecordOfFunction.ts, 9, 32))
+>Event : Symbol(Event, Decl(eventEmitterPatternWithRecordOfFunction.ts, 9, 9))
+>args : Symbol(args, Decl(eventEmitterPatternWithRecordOfFunction.ts, 9, 45))
+>Args : Symbol(Args, Decl(eventEmitterPatternWithRecordOfFunction.ts, 2, 1))
+>M : Symbol(M, Decl(eventEmitterPatternWithRecordOfFunction.ts, 8, 12))
+>Event : Symbol(Event, Decl(eventEmitterPatternWithRecordOfFunction.ts, 9, 9))
+}

--- a/tests/baselines/reference/eventEmitterPatternWithRecordOfFunction.types
+++ b/tests/baselines/reference/eventEmitterPatternWithRecordOfFunction.types
@@ -1,0 +1,21 @@
+=== tests/cases/compiler/eventEmitterPatternWithRecordOfFunction.ts ===
+interface A {
+    emit(event: string, ...args: any[]): boolean;
+>emit : (event: string, ...args: any[]) => boolean
+>event : string
+>args : any[]
+}
+
+type Args<F> = F extends (...args: infer A) => void ? A : never;
+>Args : Args<F>
+>args : A
+
+type EventMap = Record<string, Function>;
+>EventMap : Record<string, Function>
+
+interface B<M extends EventMap> extends A {
+    emit<Event extends keyof M>(event: Event, ...args: Args<M[Event]>): boolean;
+>emit : <Event extends keyof M>(event: Event, ...args: Args<M[Event]>) => boolean
+>event : Event
+>args : Args<M[Event]>
+}

--- a/tests/cases/compiler/eventEmitterPatternWithRecordOfFunction.ts
+++ b/tests/cases/compiler/eventEmitterPatternWithRecordOfFunction.ts
@@ -1,0 +1,12 @@
+// @strict: true
+interface A {
+    emit(event: string, ...args: any[]): boolean;
+}
+
+type Args<F> = F extends (...args: infer A) => void ? A : never;
+
+type EventMap = Record<string, Function>;
+
+interface B<M extends EventMap> extends A {
+    emit<Event extends keyof M>(event: Event, ...args: Args<M[Event]>): boolean;
+}

